### ISSUE TITLE
incorrect "chart validation failed" errors

### ISSUE THREAD

**Issue by bergerx** [Praqma/helmsman#507]
TL;DR: Time to time helmsman keeps failing with a clearly incorrect "chart validation failed" error for a random chart in the DSF file on consecutive runs when running on a certain environment (happens on both mac, linux with v3.3.0). I installed mitmproxy to investigate the traffic, and verified that there is no network interference, reported charts exist and being reported to exist in the queries to the helm repos by the helmsman. So the the issue seem to be in helmsman.

Details:
We have been using helmsman for a long while in my company. For certain set of DSFs it keeps failing when we are running it on our laptops, it's quite strange that this problem doesn't happen when running in the cloud.

Here is an example error reported by helmsman:
```
2020-08-19 20:54:35 INFO: Validating charts...
... {omitted irrelevant charts} ...
2020-08-19 20:54:35 DEBUG: helm search repo stable/nginx-ingress --version 1.41.2 -l
2020-08-19 20:54:35 DEBUG: helm search repo stable/prometheus-operator --version 8.5.15 -l
2020-08-19 20:54:36 DEBUG: helm search repo vmware-tanzu/velero --version 2.12.0 -l
2020-08-19 20:55:42 ERROR: Chart [ stable/prometheus-operator ] with version [ 8.5.15 ] is specified for app [rbac-manager] but was not found. If this is not a local chart, define its helm repo in the helmRepo stanza in your DSF.
2020-08-19 20:55:42 ERROR: Chart [ vmware-tanzu/velero ] with version [ 2.12.0 ] is specified for app [velero-backup] but was not found. If this is not a local chart, define its helm repo in the helmRepo stanza in your DSF.
2020-08-19 20:55:42 CRITICAL: chart validation failed

...  BUT THEY EXISTS:
# helm search repo stable/prometheus-operator --version 8.5.15 -l
NAME                      	CHART VERSION	APP VERSION	DESCRIPTION                                       
stable/prometheus-operator	8.5.15       	0.34.0     	Provides easy monitoring definitions for Kubern...
# helm search repo vmware-tanzu/velero --version 2.12.0 -l
NAME               	CHART VERSION	APP VERSION	DESCRIPTION            
vmware-tanzu/velero	2.12.0       	1.4.0      	A Helm chart for velero
# 
```

Here is a retry just with the exact same DSF:
```
2020-08-20 00:04:27 INFO: Validating charts...
... {omitted irrelevant charts} ...
2020-08-20 00:04:27 DEBUG: helm search repo stable/nginx-ingress --version 1.41.2 -l
2020-08-20 00:04:28 DEBUG: helm search repo stable/prometheus-operator --version 8.5.15 -l
2020-08-20 00:04:28 DEBUG: helm search repo vmware-tanzu/velero --version 2.12.0 -l
2020-08-20 00:05:31 ERROR: Chart [ stable/nginx-ingress ] with version [ 1.41.2 ] is specified for app [nginx-ingress] but was not found. If this is not a local chart, define its helm repo in the helmRepo stanza in your DSF.
2020-08-20 00:05:31 CRITICAL: chart validation failed

... AGAIN THE REPORTED CHART EXISTS:
# helm search repo stable/nginx-ingress --version 1.41.2 -l
NAME                	CHART VERSION	APP VERSION	DESCRIPTION                                       
stable/nginx-ingress	1.41.2       	v0.34.1    	An nginx Ingress controller that uses ConfigMap...
# 
```

The next run will likely fail on some other random chart.

I have been blaming our IT department for doing fancy stuff with our laptop traffic. After getting hit by this issue enough I decided to MITM the whole traffic and see if the `helm search repo` commands are returning something different when they run consecutively and I installed a mitmproxy to investigate all the traffic including HTTPS, verified that the issue is within helmsman's verification.

**Comment by bergerx** [comment]
Just verified that this issue still happens with v3.4.4

**Comment by luisdavim** [comment]
hi @bergerx do you have info from the `mitmproxy` regarding what changes between consecutive runs? is the payload returned by `helm search repo` different? do you have any examples you can share?

Thanks.

**Comment by bergerx** [comment]
tl; dr: 
> hi @bergerx do you have info from the mitmproxy regarding what changes between consecutive runs?

No.

> is the payload returned by helm search repo different?

No.

> do you have any examples you can share?

I'll provide one.

Some insights:

Apparently i didnt keep any output of my tests and it was not straight forward to setup the mitm environment.

But based on the network traffic all calls to `helm search repo` runs were successful as i see them on the wire, i checked all results from the helm repos one by one.

The way i checked was based on network traffic, not based on helm exec calls. And now i think about it, the tests would be simpler with a helm wrapper that calls actual helm binary and logs the invocation and result details. We can create a helm wrapper, craft a dummy dsf file with 20+ apps, each with a different chart, then keep firing helmsman. If its a race condition within helmsman (which i believe so, maybe its just some maps used in the related code path and they are not thread safe) we'll likely catch the culprit. I'll likely do this next time the issue driving me crazy.

**Comment by vpartington** [comment]
I've seen the same issue with Helmsman DSF files with many applications in them. Not on my laptop (macOS) or that of my colleague (Linux), but I have seen the issue on our Gitlab runners in Kubernetes, running `docker.io/praqma/helmsman:v3.4.6`.

I was able to resolve the issue by limiting the number of concurrent invocations of `helm search repo` to 1 by changing the value of `resourcePool` to 1:
https://github.com/Praqma/helmsman/blob/master/internal/app/main.go#L12

My guess is that Helm messes up its cache if invoked concurrently. I can imagine that this fix makes chart validation slower, but it does succeed.

My suggestion would be to remove the parallelisation code altogether as that will makes the code base simpler. Alternatively, it would be great is the amount of parallelism were configurable.

I can help by providing a PR. Please let me know!

**Comment by luisdavim** [comment]
We should probably make that `resourcePool` configurable or just use the value passed by `-p`/`--parallel`.
anyway, the latest version of helmsman does include a retry mechanism that should improve this, can you give that a try?

**Comment by luisdavim** [comment]
Can you test a build from master and let me know if you're still able to reproduce this?

**Comment by bergerx** [comment]
this is not happening anymore in our environments
